[data] [docs] Update dataset docs to reflect strict mode simplifications (#34880)
ray-project/ray: doc/source/data/transforming-data.rst

.. _transforming_data:

=================
Transforming Data
=================

Datastream transforms take in datastreams and produce new datastreams. For example, *map_batches*
is a transform that applies a
:ref:`user-defined function <transform_datastreams_writing_udfs>` on each data record
and returns a new datastream as the result. Datastream transforms can be composed to
express a chain of computations.

--------
Overview
--------

> There are two main types of supported transforms, summarized in the below table:

Reviewer: ```suggestion
There are two main types of supported transforms:
```

* One-to-one: each input block will contribute to only one output
  block, such as :meth:`ds.map_batches() <ray.data.Datastream.map_batches>`.
* All-to-all: input blocks can contribute to multiple output blocks,

>   such as :meth:`ds.random_shuffle() <ray.data.Datastream.random_shuffle>`.

Reviewer: I feel like we should move this later in the guide or remove it altogether. I'm not sure if the reader actually needs to know this

Author: Agreed.

.. list-table:: Common Ray Data transforms.
   :header-rows: 1

   * - Transform
     - Type
     - Description
   * - :meth:`ds.map() <ray.data.Datastream.map>`

>      - One-to-one

Reviewer: Same here. Not sure if it's important to know whether a transformation is one-to-one or all-to-all

Author: Agreed. I'm just leaving this to avoid making too large of a change for now.

     - Apply a given function to individual data records.
   * - :meth:`ds.map_batches() <ray.data.Datastream.map_batches>`
     - One-to-one
     - Apply a given function to batches of records.
   * - :meth:`ds.repartition() <ray.data.Datastream.repartition>`
     - All-to-all
     - | Repartition the datastream into N blocks.
   * - :meth:`ds.random_shuffle() <ray.data.Datastream.random_shuffle>`
     - All-to-all
     - | Randomly shuffle the datastream.
   * -  :meth:`ds.groupby().\<agg\>() <ray.data.Datastream.groupby>`
     - All-to-all
     - | Group data by column and aggregate each group.
   * -  :meth:`ds.groupby().map_groups() <ray.data.grouped_data.GroupedData.map_groups>`
     - All-to-all
     - | Group data by column and transform each group.

.. _transform_datastreams_writing_udfs:

--------------
Map transforms
--------------

Use ``map_batches`` to efficiently transform records in batches, or ``map`` to transform records individually:

.. tab-set::

    .. tab-item:: Map Batches
        

>       Records can be transformed in batches of ``Dict[str, np.ndarray]`` using the ``map_batches`` function. The below example shows how to use ``map_batches`` to convert text records to lowercase:

Reviewer: ```suggestion
      Call `map_batches`` to transform batches of records. Each batch has type``Dict[str, np.ndarray]``. The below example shows how to use ``map_batches`` to convert text records to lowercase:
```

      .. literalinclude:: ./doc_code/transforming_data.py
        :language: python
        :start-after: __map_batches_begin__
        :end-before: __map_batches_end__

    .. tab-item:: Map

>        Records can also be transformed one at a time using the ``map`` function, which takes records encoded as ``Dict[str, Any]]``. The below example shows how to convert text records to lowercase:

Reviewer: Also active voice here

       .. literalinclude:: ./doc_code/transforming_data.py
         :language: python
         :start-after: __map_begin__
         :end-before: __map_end__

Configuring resources
=====================

> By default, each task used for  (e.g., `map` or `map_batches`) will request 1 CPU from Ray.

Reviewer: ```suggestion
By default, each task used for  (e.g., `map` or `map_batches`) requests 1 CPU from Ray.
```

      values represent a batch of record field values.

      .. literalinclude:: ./doc_code/transforming_data.py
        :language: python
        :start-after: __writing_numpy_udfs_begin__
        :end-before: __writing_numpy_udfs_end__

    .. tab-item:: Pandas

      The ``"pandas"`` batch format presents batches in
      `pandas.DataFrame <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`__
      format.

      .. literalinclude:: ./doc_code/transforming_data.py
        :language: python
        :start-after: __writing_pandas_udfs_begin__
        :end-before: __writing_pandas_udfs_end__

    .. tab-item:: PyArrow

      The ``"pyarrow"`` batch format presents batches in
      `pyarrow.Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__
      format.

      .. literalinclude:: ./doc_code/transforming_data.py
        :language: python
        :start-after: __writing_arrow_udfs_begin__
        :end-before: __writing_arrow_udfs_end__

.. _transforming_data_actors:

Reduce setup overheads using actors
===================================

Data transforms can be executed by either :ref:`Ray tasks <ray-remote-functions>`

> or :ref:`Ray actors <actor-guide>`. By default, tasks are
> used. For transforms that require expensive setup,

Reviewer: Active voice
```suggestion
or :ref:`Ray actors <actor-guide>`. By default,``map_batches`` uses tasks.
```
